# Acropora_Mutation_Vasquez_Kitchen_et_al.

Here we provide the code associated with the bioRxiv preprint, *"Juvenile corals inherit mutations acquired during the parentâ€™s lifespan"*, by Vasquez Kuntz and Kitchen et al. 2020 (https://www.biorxiv.org/content/10.1101/2020.10.19.345538v1).

The raw Affymetrix CEL files for the parent *A. palmata* colony, neighboring *A. palmata* colonies and offspring generated in this study were processed through the Standard Tools for Acroporid Genotyping pipeline (https://coralsnp.science.psu.edu/galaxy/) as previously reported (Kitchen et al. 2020, Scientific Reports; https://www.nature.com/articles/s41598-020-69101-z). During this process, the genotypes are converted to a variant call format (VCF) file and merged with other database samples at the time (n = 513 in total). The VCF files is available at: https://scholarsphere.psu.edu/resources/76e0ccdd-fbe0-4190-92cf-2a91e02bd73b.

To identify the relationship between the parent and offsprings, genotypes were extracted to calculate pairwise genetic distances, percent heterozygosity and recombination breaks of the offspring (Genetic_Dist_Het_Recombination_array.R). Potential mutations were identified from the genotypes without missing data from all parental samples and neighboring colonies, which were found to be clone mates (shared multilocus genotype). (extractGentoype_idMutations.R)

The B-allele frequency and logR ratio, two values of array fluorescent signal intensity, were calculated as part of the STAG pipeline using the *affy2vcf* bcftools plugin (https://github.com/freeseek/gtc2vcf). These values were extracted and visualized for each reference scaffold. (BAF_LRR_plots_per_scaffold.R) From the BAF and LRR values, deviations from signal intensities expected for 2 allele copies (a diploid sample) were assessed using the CGHcall R package to identify potential copy number variants or genetic mosaics that might appear as post-embyronic mutations (CNV_analysis.R). 
